File: server/src/core/sensor/sensor_manager.py
from abc import ABCMeta, abstractmethod


class SensorManager(object):
    """
    API for handling Sensors
    """
    __metaclass__ = ABCMeta

    @abstractmethod
    def add_sensor(self, sensor_id, sensor):
        """
        Add a new sensor
        :param sensor_id: the sensor_id
        :param sensor: the sensor to add
        :raise SensorAlreadyExistsException: if the sensor was already added
        :return: the sensor added
        """
        raise NotImplementedError

    @abstractmethod
    def get_sensor(self, sensor_id):
        """
        Get a sensor by id
        :param sensor_id: the unique id of the sensor to get
        :raise UnknownSensorException: if a sensor with that id does not exist
        :return: the sensor retrieved
        """
        raise NotImplementedError

    @abstractmethod
    def remove_sensor(self, sensor_id):
        """
        Remove an sensor by id
        :param sensor_id: The id to uniquely locate the sensor to remove
        :raise: UnkownSensorException: If the sensor is not found
        :return: The sensor with the given id
        """
        raise NotImplementedError

    @abstractmethod
    def update_sensor(self, sensor_id, sensor):
        """
        Update an already existent sensor.
        :param sensor_id: The id of the sensor to be updated
        :param sensor: The new sensor that will replace the old one with new information
        :raise: UnknownSensorException if no sensor is found with the given id.
        :return: The new sensor updated
        """
        raise NotImplementedError

    @abstractmethod
    def get_all_sensors(self):
        """
        Return all registered sensors
        :return: all sensors
        """
        raise NotImplementedError



class SensorManagerException(Exception):
    """
    Root exception related to an SensorManager
    """


class SensorAlreadyExistsException(SensorManagerException):
    """
    Throw this exception when wanting to add a sensor that already exists
    """
    pass


class UnknownSensorException(SensorManagerException):
    """
    Throw this exception when the requested sensor is not found
    """
    pass

File: android-client/location-api/api/src/main/java/com/example/location/api/system/SensorManager.java
package com.example.location.api.system;

import com.example.location.api.entity.sensor.Sensor;
import com.example.location.api.entity.sensor.SensorConfiguration;

/**
 * Handles {@link Sensor}s.
 */
public interface SensorManager {

    /**
     * Try to get a sensor with the given id. If it does not exists,
     * create a new {@link Sensor} using the given and register it in the system
     * @param config The sensor configuration to be created
     * @return The new Sensor being created
     */
    Sensor getOrCreateSensor(String sensorId, SensorConfiguration config) throws SensorManagerException;

    /**
     * Create a new {@link Sensor} and register it in the system.
     * If the sensor already exists, throw an exception.
     * @param config The sensor configuration to be created
     * @return The new Sensor being created
     */
    Sensor createSensor(SensorConfiguration config) throws SensorManagerException;

}
